### PR TITLE
Feat/mss util

### PR DESCRIPTION
**Pull request by paulhandy**
Added two utils: one to generate Curl-P-81 iss v2 merkle tree (as mam uses), and one to sign (v2) a message (must be padded to 81 trytes) whose digest is generated with Kerl.

### CODE REVIEW

File: utils/signer/mssign.c [hunk 1]
```diff
@@ -0,0 +1,183 @@
+#include "common/curl-p/trit.h"
+#include "common/kerl/hash.h"
+#include "common/sign/v2/iss_curl.h"
+#include "common/trinary/trit_tryte.h"
+#include "mam/v1/merkle.h"
```
Comment (thibault-martinez) [inline utils/signer/mssign.c:5]: Maybe mam/v1 is not the right place for the merkle tree anymore ?

File: utils/signer/mssign.c [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+  if (security == 2 && sum != 0) {
+    return 1;
+  }
+  for (i = 2 * HASH_LENGTH / 3; i < HASH_LENGTH; i++) {
+    sum += hash[i];
+  }
+  return sum != 0;
+}
+
+int increment_trits(trit_t *t, size_t length) {
```
Comment (thibault-martinez) [inline utils/signer/mssign.c:56]: Don't we have that already somewhere ?

File: utils/signer/mssign.c [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+  size_t sec;
+  trit_t *msg;
+  size_t msg_len;
+  trit_t seed[HASH_LENGTH];
+  trit_t *tree;
+  size_t tree_len;
+  size_t n_leafs;
+} ToSign;
+
+int test_hash(trit_t *hash, size_t security) {
```
Comment (thibault-martinez) [inline utils/signer/mssign.c:34]: You can use the existing `signed_window` function https://github.com/iotaledger/entangled/blob/develop/common/sign/v2/iss.c.inc#L78

File: utils/signer/mssign.c [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+    trit_t sigkey[key_length];
+    tryte_t signature[(key_length / 3) + 1];
+    trit_t hash[HASH_LENGTH] = {0};
+    find_nonce(args, hash);
+
+    memset(signature, 0, sizeof(signature));
+
+    iss_curl_subseed(args->seed, args->seed, args->key_i, &c);
+    iss_curl_key(args->seed, sigkey, key_length, &c);
+    curl_reset(&c);
```
Comment (thibault-martinez) [inline utils/signer/mssign.c:130]: `iss_curl_key` is already reseting

File: utils/signer/mssign.c [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+    iss_curl_signature(sigkey, hash, 0, sigkey, key_length, &c);
+    curl_reset(&c);
```
Comment (thibault-martinez) [inline utils/signer/mssign.c:132]: `iss_curl_signature`is already reseting

File: utils/signer/mssign.c [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+  } while (test_hash(hash, args->sec));
+  trits_to_trytes(nonce, nonce_str, sizeof(nonce) / sizeof(trit_t));
+  fprintf(stdout, "%s\n", nonce_str);
+  return 0;
+}
+
+int leaf_count(size_t count, size_t depth, size_t acc) {
+  size_t acc_next = acc + (1 << depth);
+  if (acc_next >= count) {
+    return 1 << depth;
```
Comment (thibault-martinez) [inline utils/signer/mssign.c:91]: Are we assuming a perfect / balanced tree ?

File: utils/signer/mssign.c [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+  init_curl(&c);
+  {
+    trit_t sigkey[key_length];
+    tryte_t signature[(key_length / 3) + 1];
+    trit_t hash[HASH_LENGTH] = {0};
+    find_nonce(args, hash);
+
+    memset(signature, 0, sizeof(signature));
+
+    iss_curl_subseed(args->seed, args->seed, args->key_i, &c);
```
Comment (thibault-martinez) [inline utils/signer/mssign.c:128]: Shouldn't you add up the offset to the key index ?

File: utils/signer/mssign.c [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+      case 'o':
+        field.offset = strtol(optarg, (char **)NULL, 10);
+        break;
+    }
+  }
+  if (field.sec == 0) {
+    field.sec = 1;
+  }
+  if (field.key_i >= field.n_leafs) {
+    printf("key index outside of range\n");
```
Comment (thibault-martinez) [inline utils/signer/mssign.c:179]: Not always true given `field.offset`

File: utils/signer/mssverify.c [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+    {
+      size_t sig_len = strlen((char *)args->sig);
+      trit_t sig[sig_len * 3];
+
+      memset(sig, 0, sizeof(sig));
+
+      trytes_to_trits(args->sig, sig, sig_len);
+      iss_curl_sig_digest(hash, hash, 0, sig, sig_len * 3, &c);
+      iss_curl_address(hash, hash, HASH_LENGTH, &c);
+      curl_reset(&c);
```
Comment (thibault-martinez) [inline utils/signer/mssverify.c:64]: `iss_curl_address` is already reseting

File: utils/signer/mssign.c [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+    return 1;
+  }
+  for (i = 2 * HASH_LENGTH / 3; i < HASH_LENGTH; i++) {
+    sum += hash[i];
+  }
+  return sum != 0;
+}
+
+int increment_trits(trit_t *t, size_t length) {
+  trit_t *end = &t[length];
```
Comment (jserv) [inline utils/signer/mssign.c:57]: Why not simply using `trit_t *end = t + length;`?

File: utils/signer/mssign.c [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+#include "common/sign/v2/iss_curl.h"
+#include "common/trinary/trit_tryte.h"
+#include "mam/v1/merkle.h"
+
+#include <getopt.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+
+#define CHARGS "d:s:i:x:t:"
```
Comment (thibault-martinez) [inline utils/signer/mssign.c:12]: 'd' should be 'm'

File: utils/signer/mssign.c [hunk 12]
```diff
@@ -0,0 +1,183 @@
...
+  }
+  return t == end;
+}
+
+int find_nonce(ToSign *args, trit_t *hash) {
+  trit_t nonce[HASH_LENGTH] = {0};
+  tryte_t nonce_str[(HASH_LENGTH / 9) + 1] = {0};
+  Kerl k, k_copy;
+  init_kerl(&k);
+  kerl_absorb(&k, args->msg, args->msg_len);
```
Comment (thibault-martinez) [inline utils/signer/mssign.c:76]: This will assert if `args->msg_len % HASH_LENGTH != 0`, do we assume the message has this kind of length ?

**Pull request closed without merging**